### PR TITLE
[com_fields] Reload task belongs to the component controller

### PR DESCRIPTION
**Pull request by laoneo**
Pull Request for Issue #15995.

### Summary of Changes
When a category changes for an article the page got reloaded to determine the custom fields. Actually this action got delegated to the com_fields controller. On that point the data is unvalidated stored in the session. This can lead to time shifts when the Joomla timezone is not UTC, more information can be found in issue #15995.

This pr reloads the page trough the component controller and not the com_fields one.

### Testing Instructions
- Set the global Joomla timezone to a different one than UTC, for example Berlin
- Create two article categories
- Create a calendar custom field
- Set show time to yes in the field
- Save the field
- Create an article
- Change the category of the article

### Expected result
The time in the custom field remains.

### Actual result
The time in the custom field shifts the same hours as the Joomla timezone has.

### CODE REVIEW

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on 64d9bb2ff51aba0b6d7bd40a688de3301a5e865b<br /><br />Set Time-Zone to "Berlin", 
applied PR, 
change Category > +2h in Calendar-Field
changed back to 1st Category > additional 2h (4 Hours)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

**Comment by laoneo** [comment]
Do the publishing dates also shift?

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 64d9bb2ff51aba0b6d7bd40a688de3301a5e865b<br /><br />Test again, all works fine now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

**Comment by wilsonge** [comment]
We need to be careful of ACL here - this is going to expose public data if accessed directly from a URL

**Comment by laoneo** [comment]
Really, it uses the same code and checks as the save function. Which line do you think can be dangerous?

**Comment by Bakual** [comment]
> Really, it uses the same code and checks as the save function. Which line do you think can be dangerous?

Maybe I miss something but I don't see any ACL checks at all here.
The save method uses `$this->allowSave($data, $key)` to check if saving is allowed which I don't see in your reload method.
We may also have to use `checkEditId($context, $id)` to verify that the item was correctly opened before.

**Comment by laoneo** [comment]
Why is the allowSave check needed, it doesn't save, it just reloads the page to an url like /cms/administrator/index.php?option=com_content&view=article&layout=edit&id=1. The checkedit id is done in the controller itself. I mean at the end it performs the same action as when you want to edit the article. If you paste that url directly into the browser and you didn't go trough the edit task, then it kicks you back to the list with a warning to not get directly to the edit page. I'v just tested it and it worked as expected.
If I'm now not completely wrong then I don't understand your points here. What else needs to be checked?

**Comment by Bakual** [comment]
> Why is the allowSave check needed, it doesn't save, it just reloads the page to an url like /cms/administrator/index.php?option=com_content&view=article&layout=edit&id=1.

It does save the data into the session. It at least feels wrong to have a controller task without any ACL check. Which means anyone could run it. It may not be an issue by itself, but it may bite us back and there is no reason this task should be available unauthorised. Using allowSave or allowEdit (if that fits better due to not saving into db) is a save check we can and should do imho. I don't see a reason to not do it.

As for checkEditId, you're right. I just remembered we check it but didn't remember where exactly :smile:

**Comment by wilsonge** [comment]
Does this not allow accessing data from any JControllerForm instance (including in the frontend (e.g. com_contact)) without checking the published/acl levels for the item though?

**Comment by laoneo** [comment]
I don't se how, it only stores the jform data into the session.

**Comment by mbabker** [comment]
It is still a POST request receiving user data.  It should therefore be properly validated and access controlled.  Especially if that data stored into the session can be used in a follow on request (which HAS been used in security issues we have fixed in the last 2 years).

**Comment by wilsonge** [comment]
The session is data that the user controls. That means the user has the data which they shouldn't have been allowed to access (for ACL reasons) - and even if it isn't displayed on the page. It's trivial to get the session data if you wanted it.

**Comment by laoneo** [comment]
Does that help https://github.com/joomla/joomla-cms/pull/16665/commits/d10e8fc78bada54bdd6b72feffbb0c6b4c4420d5?

**Comment by Bakual** [comment]
Imho, the ACL check is now fine.

Looking at the redirects, why don't you just use `$this->redirect()`? Also `$app->close()` is useless even when you use `$app->redirect()` directly, since that method already closes the app. Or do I miss something?

**Comment by laoneo** [comment]
Just to be sure. Probably not needed.

**Comment by Bakual** [comment]
If there is no specific reason to use `$app->redirect()` directly, can you just use `$this->setRedirect($url, $msg, $type)` and `$this->redirect()` then? So we are consistent with how we process redirects in other places in the controllers.

**Comment by laoneo** [comment]
True, done.

**Comment by wilsonge** [comment]
RE ACL: I'm honestly unsure given what this function does whether it should be edit or read acl permissions we should be using. But this looks significantly better now :)

**Comment by laoneo** [comment]
It should check edit, as it belongs to the edit action.

**Comment by infograf768** [comment]
@laoneo 
I can't get the calendar field anymore
( ! ) Warning: Invalid argument supplied for foreach() in /Applications/MAMP/htdocs/testnew/trunkgitnew/libraries/cms/helper/content.php on line 74

**Comment by infograf768** [comment]
fields

**Comment by infograf768** [comment]
Forgot to delete `administrator/components/com_fields/controllers/field.php`

After deleting it, I get:
`An error has occurred.`
` 0 Invalid controller: name='field', format='' `

File: administrator/components/com_admin/script.php [hunk 1]
```diff
@@ -1723,6 +1723,8 @@ public function deleteUnexistingFiles()
 			'/administrator/templates/hathor/html/mod_menu',
 			'/administrator/components/com_messages/layouts/toolbar',
 			'/administrator/components/com_messages/layouts',
+			// Joomla! __DEPLOY_VERSION__
+			'/components/com_fields/controllers',
```
Comment (infograf768) [inline administrator/components/com_admin/script.php]: this targetting a folder. Should be the file.
`'/components/com_fields/controllers/field.php',`

Comment (laoneo) [inline administrator/components/com_admin/script.php]: This folder has only one file, so better to remove it completely.

**Comment by laoneo** [comment]
Good catch. Fixed it in the last commits.

File: administrator/components/com_admin/script.php (hunk 1 above)
Comment (infograf768) [inline administrator/components/com_admin/script.php]: better do both. see https://github.com/Digital-Peak/joomla-cms/blob/feb148b97dbab493f21a02a289ddf8d06df03692/administrator/components/com_admin/script.php#L1599

Comment (infograf768) [inline administrator/components/com_admin/script.php]: https://github.com/Digital-Peak/joomla-cms/blob/feb148b97dbab493f21a02a289ddf8d06df03692/administrator/components/com_admin/script.php#L1599
better do both

Comment (laoneo) [inline administrator/components/com_admin/script.php]: Why?

Comment (infograf768) [inline administrator/components/com_admin/script.php]: As comment says, deleting only the folder may not work when ftp is used.
It is safer to at least delete the file.

**Comment by infograf768** [comment]
Works now. Remains the necessary changes in script.php

**Comment by laoneo** [comment]
Ok, done.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on fb925e7cd3b0304cb778297cf2d08918670c9994<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on fb925e7cd3b0304cb778297cf2d08918670c9994<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16665">issues.joomla.org/tracker/joomla-cms/16665</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful tests.

**Comment by laoneo** [comment]
Thanks guys!

**Pull request merged**